c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+#include <c10/cuda/CUDACachingAllocator.h>
+#include <c10/cuda/CUDAException.h>
+#include <c10/cuda/CUDAFunctions.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/util/UniqueVoidPtr.h>
+#include <c10/util/irange.h>
+
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace c10 {
+namespace cuda {
+namespace CUDACachingAllocator {
+namespace CudaMallocAsync {
+
+#if CUDA_VERSION > 11040
+// CUDA device allocator that uses cudaMallocAsync to implement
+// the same interface as CUDACachingAllocator.cpp.
+
+// Designed to be safe for CUDA graph capture.
+// Interactions with CUDA graph capture are mediated by
+// notifyCaptureBegin
+// notifyCaptureAboutToEnd
+// notifyCaptureEnded
+// notifyCaptureDestroy
+
+// Implementation details, not declared in CUDACachingAllocator.h
+namespace {
+
+// General helpers
+
+struct UsageStream {
+  cudaStream_t stream;
+  int device;
+  UsageStream() {}
+  UsageStream(cudaStream_t s, int d) : stream(s), device(d) {}
+  UsageStream(const UsageStream& us) : stream(us.stream), device(us.device) {}
+  UsageStream(const UsageStream&& us) : stream(us.stream), device(us.device) {}
+  UsageStream& operator=(UsageStream other) {
+    stream = other.stream;
+    device = other.device;
+    return *this;
+  }
+};
+
+struct PtrUsage {
+  std::vector<UsageStream> usage_streams;
+  uint64_t size;
+  bool captured;
+  PtrUsage(uint64_t s, bool c) : size(s), captured(c) {}
+};
+
+int device_count = 0;
+// these don't need to be std::once_flags as in CUDAGeneratorImpl.cpp
+// because they'll only be flipped by functions that have locked the mutex.
+std::vector<bool> devs_initialized_flags;
+std::vector<UsageStream> dummy_unifying_free_streams;
+
+// Possible micro-optimization:
+// Some accesses to ptr_info are read-only.
+// We could let those be concurrent with a shared_mutex and
+// have concurrent calls take a shared_lock.
+// Keeping it simple with an ordinary mutex for now.
+std::mutex general_mutex;
+
+/**
+ * Note [Avoid freeing uncaptured ptrs during CUDA graph capture]
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * During CUDA graph capture, it's illegal to call cudaFreeAsync
+ * on a pointer that came from a non-captured cudaMallocAsync.
+ * Unfortunately, Python being what it is, it's impossible to be
+ * sure no uncaptured tensor will ever have its destructor called
+ * in a capturing region.
+ * We avoid errors by
+ *  1. remembering if allocated pointers were captured or uncaptured
+ *  2. during capture, if we detect an attempt to free an uncaptured
+ *     allocation on a capturing stream, don't free it immediately,
+ *     just remember it and defer its cudaFreeAsync call to after
+ *     the end of capture (specifically, to notifyCaptureEnded).
```

Comment:
It's a restriction of the CUDA API because ungraphed allocations are meant to be one-off, while graphs are meant to be replayed many times. I think their philosophy is: if cuda silently allowed ungraphed allocations to be freed during capture, an ungraphed allocation address that happened to get freed during capture would be actually-freed on the first replay, then later replays would attempt to actually-free the same address, at which point it would either no longer be live or be associated with some totally different allocation (at which point in either scenario an actual free is a bad thing to do).

Reply:
Cuda guys reconcile this "inconsistency" in their brains by regarding
> on replays it will ignore allocations recorded during graph capture

as "auto free on launch", ie, when the replay sees some captured VA that is unfreed since the last launch, it (conceptually or literally, im not sure) frees and immediately reallocs that VA.

The graph should not ignore frees of graphed memory, because the desired situation is:
```
graph1.replay(); // e.g. forward pass of a make_graphed_callable, allocates some activations
graph2.replay(); // e.g. backward pass of make_graphed_callable, consumes and frees the activations
// graph2.replay()s frees should have a real effect, because out here we shouldn't regard the activations as allocated and their PA unusable by later cudaMallocAsyncs.
```